tensorly/decomposition/mda.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from .. import backend as T
+from ..base import unfold
+from ..tenalg import multi_mode_dot
+
+
+# Author: James Oldfield
+
+# License: BSD 3 clause
+
+
+def mda(X, y, ranks, n_iters=5):
+    """Multilinear Linear Discriminant Analysis (MDA).
+
+        Learns a projection matrix for each mode of the data tensor
+        to project the input tensor into a low-dimensional tensor subspace
+        where the *scatter ratio criterion* is maximised.
+
+    Parameters
+    ----------
+    X : ndarray
+        tensor data of shape (n_samples, N1, ..., NS). Note: the first dimension is the sample dimension.
+    y : ndarray
+        list of length (n_samples) containing the integer-valued class label of tensor sample i.
+    ranks : list
+        list of integers determining the dimension of the subspace for mode-k.
+    n_iters : int, optional, default is 5
+        number of steps to employ the APP scheme for.
+
+    Returns
+    -------
+    factors : list
+        list of the learnt projection matrices for each mode
+
+    Notes
+    -----
+
+    This implementation computes the *Constrained Multilinear Discriminant Analysis* (CMDA) solution as presented in [1].
+
+    Given the learnt factor matrices, one can then compute the projection along all modes with::
+
+      factors = mda(X_train, y_train, ranks, n)
+      Z = tl.tenalg.multi_mode_dot(X_train, factors, modes=[1, 2], transpose=True)
+
+    - [1] Q. Li and D. Schonfeld, "Multilinear Discriminant Analysis for Higher-Order Tensor Data Classification," in IEEE Transactions on Pattern Analysis and Machine Intelligence, vol. 36, no. 12, pp. 2524-2537, 1 Dec. 2014, doi: 10.1109/TPAMI.2014.2342214.
+
+    """
+
+    ###############
+    # check correct # of ranks have been supplied
+    ###############
+    assert len(ranks) == len(T.shape(X)[1:]), 'Expected number of ranks: {}. \
+        But number supplied is {}'.format(len(T.shape(X)[1:]), len(ranks))
+
+    backend = T.get_backend()
+
+    global_mean = T.mean(X, axis=0)
+    class_means = []
+
+    # ith element will contain a list of the indices of training data with class label i
+    class_idx = [[] for _ in range(len(set(y)))]
+
+    # store the training data's class label at class index
+    for i, label in enumerate(y):
+        class_idx[label] += [i]
+
+    # store the mean of all tensor samples with label i
+    for i in range(len(set(y))):
+        # tensorflow is only backend to not support indexing into tensor with a list
+        if backend == 'tensorflow':
+            class_means += [T.mean(T.tensor([X[j] for j in class_idx[i]]), axis=0)]
+        else:
+            class_means += [T.mean(X[class_idx[i], ...], axis=0)]
```

Comment:
We can directly get the means without creating `class_idx`: 
```suggestion
            class_means += [T.mean(X[y == i, ...], axis=0)]
```